firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexBackfiller.java

```diff
@@ -103,13 +131,105 @@ public BackfillScheduler newScheduler(AsyncQueue asyncQueue, LocalStore localSto
     return new BackfillScheduler(asyncQueue, localStore);
   }
 
-  // TODO(indexing): Figure out which index entries to backfill.
-  public Results backfill() {
-    int numIndexesWritten = 0;
+  public Results backfill(LocalStore localStore) {
+    // TODO(indexing): Handle field indexes that are removed by the user.
+    fetchNewFieldIndexes();
+
+    int numIndexesWritten = writeIndexEntries(localStore);
     int numIndexesRemoved = 0;
     return new Results(/* hasRun= */ true, numIndexesWritten, numIndexesRemoved);
   }
 
+  private void fetchNewFieldIndexes() {
+    List<FieldIndex> fieldIndexes = indexManager.getFieldIndexes(lastFieldIndexId + 1);
+
+    if (fieldIndexes.size() > 0) {
+      fieldIndexQueue.addAll(fieldIndexes);
+
+      // Since index configuration entries are sorted by index id, we can set the last fetched id to
+      // the id of the last field index in the array.
+      lastFieldIndexId = fieldIndexes.get(fieldIndexes.size() - 1).getIndexId();
+    }
+  }
+
+  /** Writes index entries based on the FieldIndexQueue. Returns the number of entries written. */
+  private int writeIndexEntries(LocalStore localStore) {
+    int numIndexesWritten = 0;
+    SnapshotVersion lastReadTime = null;
+
+    // TODO(indexing): Track the number of entries written and process more field indexes if
+    // we're under the maximum.
+    int lastArrayIndex = Math.min(maxFieldIndexesToProcess, fieldIndexQueue.size());
+    List<FieldIndex> fieldIndexesToProcess = fieldIndexQueue.subList(0, lastArrayIndex);
+
+    // Create a map of each collection group to all the FieldIndexes on it.
+    Map<String, ArrayList<FieldIndex>> collectionToFieldIndexes = new HashMap<>();
+
+    // Map of field index to the most recent snapshot version written.
+    Map<FieldIndex, SnapshotVersion> fieldIndexToNewestVersion = new HashMap<>();
```

Comment:
Do we need to store the snapshot version separately? The FIeldIndex already has a snapshot version which we could just update. This map would then likely change to `Map<IndexId, FieldIndex>`.

Reply:
Done. We can remove the map completely in this case since we already have the list of field indexes that we processed here.